File: 2023/src/day22.zig
const std = @import("std");
const ascii = @import("std").ascii;
const utils = @import("./utils.zig");

const fileData = @embedFile("./files/day22.txt");
const fileDataEx = @embedFile("./files/day22ex.txt");

const Regex = @import("zig-regex").Regex;

const Allocator = std.mem.Allocator;
fn line_intersect(comptime T: type, x1: T, y1: T, x2: T, y2: T) bool {
    if (x1 >= x2 and y1 <= y2) {
        return true;
    }
    if (x2 >= x1 and y2 <= y1) {
        return true;
    }
    if (x1 <= y2 and y1 >= y2) {
        return true;
    }
    if (x2 <= y1 and y2 >= y1) {
        return true;
    }
    return false;
}
const Cube = struct {
    name: []const u8,
    supports: std.ArrayList([]const u8),
    p1: [3]u32,
    p2: [3]u32,
    is_supported_by: std.StringHashMap(bool),
    is_supported_by_tmp: std.StringHashMap(bool),

    pub fn init(allocator: Allocator, name: []const u8) Cube {
        return Cube{ .name = name, .p1 = [3]u32{ 0, 0, 0 }, .p2 = [3]u32{ 0, 0, 0 }, .supports = std.ArrayList([]const u8).init(allocator), .is_supported_by = std.StringHashMap(bool).init(allocator), .is_supported_by_tmp = std.StringHashMap(bool).init(allocator) };
    }

    pub fn intersect(self: Cube, other: Cube) bool {
        return line_intersect(u32, self.p1[0], self.p2[0], other.p1[0], other.p2[0]) and line_intersect(u32, self.p1[1], self.p2[1], other.p1[1], other.p2[1]);
    }

    pub fn reset(self: *Cube) !void {
        self.is_supported_by_tmp.clearAndFree();
        var it = self.is_supported_by.iterator();
        while (it.next()) |obj| {
            try self.is_supported_by_tmp.put(obj.key_ptr.*, false);
        }
    }

    pub fn addToGround(self: *Cube, cubes: []Cube) void {
        var zmin: u32 = 1;
        for (cubes) |cube| {
            if (self.intersect(cube)) {
                zmin = @max(zmin, @max(cube.p1[2], cube.p2[2]) + 1);
            }
        }
        var offset = @min(self.p1[2], self.p2[2]);
        self.p1[2] = self.p1[2] - offset + zmin;
        self.p2[2] = self.p2[2] - offset + zmin;
    }

    pub fn fill_supports(self: *Cube, cubes: []Cube) !void {
        for (cubes) |*cube| {
            if (cube == self) {
                continue;
            }
            if ((self.p2[2] == cube.p1[2] - 1) and self.intersect(cube.*)) {
                try self.supports.append(cube.name);
                try cube.is_supported_by.put(self.name, false);
            }
        }
    }
};

fn reset_cubes(
    cubes: *std.StringHashMap(Cube),
) !void {
    var it = cubes.iterator();
    while (it.next()) |obj| {
        try obj.value_ptr.*.reset();
    }
}

fn printStringHashMap(data: std.StringHashMap(bool)) void {
    var it = data.iterator();
    while (it.next()) |item| {
        std.debug.print("{s},", .{item.key_ptr.*});
    }
}

// should return true if a < b :)
fn cmpFunc(context: void, a: Cube, b: Cube) bool {
    _ = context;
    var a_z = @min(a.p1[2], a.p2[2]);
    var b_z = @min(b.p1[2], b.p2[2]);
    return a_z < b_z;
}

fn parse_cube(
    line: []const u8,
    counter: usize,
    allocator: Allocator,
    cubes: *std.ArrayList(Cube),
) !void {
    var re = try Regex.compile(allocator, "(\\d+),(\\d+),(\\d+)~(\\d+),(\\d+),(\\d+)");
    var captures = try re.captures(line);
    if (captures != null) {
        var cube = Cube.init(allocator, try std.fmt.allocPrint(allocator, "{}", .{counter}));
        for (0..3) |ix| {
            cube.p1[ix] = try std.fmt.parseInt(u32, captures.?.sliceAt(ix + 1).?, 10);
            cube.p2[ix] = try std.fmt.parseInt(u32, captures.?.sliceAt(ix + 4).?, 10);
        }
        try cubes.append(cube);
    } else {
        std.debug.print("{s} not found\n", .{line});
        unreachable;
    }
}

fn chain_reaction(cubes: *std.StringHashMap(Cube), allocator: Allocator, first: []const u8) !i32 {
    var counter: i32 = 0;
    const QueueType = std.ArrayList([]const u8);
    var queue = QueueType.init(allocator);
    try queue.append(first);

    while (queue.items.len > 0) {
        var item = queue.orderedRemove(0);
        counter += 1;
        var cube = cubes.get(item).?;
        for (cube.supports.items) |supported| {
            var supported_cube = cubes.getPtr(supported).?;
            _ = supported_cube.is_supported_by_tmp.remove(item);
            if (supported_cube.is_supported_by_tmp.count() == 0) {
                try queue.append(supported_cube.name);
            }
        }
    }
    return counter - 1;
}

fn solve(data: anytype) !void {
    var arena_state = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    defer arena_state.deinit();
    const allocator = arena_state.allocator();

    var splitsLines = std.mem.split(u8, data, "\n");

    var cubes = std.ArrayList(Cube).init(allocator);
    var counter: usize = 0;
    while (splitsLines.next()) |l| : (counter += 1) {
        try parse_cube(l, counter, allocator, &cubes);
    }

    std.debug.print("{} cubes found \n", .{cubes.items.len});
    std.sort.insertion(Cube, cubes.items, comptime {}, comptime cmpFunc);

    // Add to ground
    for (cubes.items, 0..) |*cube, ix| {
        cube.addToGround(cubes.items[0..ix]);
    }
    std.sort.insertion(Cube, cubes.items, comptime {}, comptime cmpFunc);

    for (cubes.items) |*cube| {
        try cube.fill_supports(cubes.items);
    }

    // for (cubes.items) |cube| {
    //     std.debug.print("Cube {s}: {any} -> {any} is supported by {}, supports {}\n", .{ cube.name, cube.p1, cube.p2, cube.is_supported_by.count(), cube.supports.count() });
    // }

    // Part 1
    var only_support = std.StringHashMap(bool).init(allocator);
    for (cubes.items) |cube| {
        if (cube.is_supported_by.count() == 1) {
            var it = cube.is_supported_by.iterator();
            try only_support.put(it.next().?.key_ptr.*, false);
        }
    }
    var could_be_removed = cubes.items.len - only_support.count();
    std.debug.print("Part 1: {} cubes could be dissintegrated\n", .{could_be_removed});

    // Pärt 2
    var cube_hash = std.StringHashMap(Cube).init(allocator);
    for (cubes.items) |cube| {
        try cube_hash.put(cube.name, cube);
    }

    var part2: i32 = 0;
    for (cubes.items) |cube| {
        try reset_cubes(&cube_hash);
        var fallen = try chain_reaction(&cube_hash, allocator, cube.name);
        //std.debug.print("Removing {s} -> {} cube falls \n", .{ cube.name, fallen });
        part2 += fallen;
    }
    std.debug.print("Part 2: {} \n", .{part2});
}

pub fn main() !void {
    const argv = std.os.argv;
    var example: bool = false;
    std.log.info("Hello day 22: {s}", .{argv});
    for (argv) |args| {
        if (std.mem.eql(u8, args[0..2], "ex")) {
            example = true;
        }
    }

    try solve(if (example) fileDataEx else fileData);
}

File: 2023/src/day22.py
import argparse
import queue
import copy
import re
import math
from collections import deque


def line_intersect(a, b):
    if a[0] >= b[0] and a[1] <= b[1]:
        return True
    if b[0] >= a[0] and b[1] <= a[1]:
        return True
    if a[0] <= b[1] and a[1] >= b[1]:
        return True
    if b[0] <= a[1] and b[1] >= a[1]:
        return True
    # if a[0] <= b[0] and a[1] >= b[0]:
    #     return True
    return False


class Cube:
    def __init__(self, line, name):
        a, b = line.strip().split("~")
        self.name = name
        self.x1, self.y1, self.z1 = list(map(int, a.split(",")))
        self.x2, self.y2, self.z2 = list(map(int, b.split(",")))
        if self.x1 > self.x2 or self.y1 > self.y2 or self.z1 > self.z2:
            print("Should sort")
        self.supports = []
        self.is_supported_by = []

    def intersect(self, other):
        return line_intersect((self.x1, self.x2), (other.x1, other.x2)) and line_intersect((self.y1, self.y2), (other.y1, other.y2))
        # return self.x1 < other.x2 and self.x2 > other.x1 and self.y1 < other.y2 and self.y2 > other.y1

    def add_to_ground(self, ground_map):
        z_min = 1
        for obj in ground_map:
            if self.intersect(obj):
                z_min = max(z_min, max(obj.z1, obj.z2)+1)
                # print(f"{self.name} intersect {obj.name}, z min is {z_min}")

        offset = min(self.z1, self.z2)
        self.z1 = self.z1 - offset + z_min
        self.z2 = self.z2 - offset + z_min

    def count_supports(self, ground_map):
        for obj in ground_map:
            if obj == self:
                continue
            # if z in contact:
            if self.z2 == (obj.z1-1):
                if self.intersect(obj):
                    self.supports.append(obj.name)
                    obj.is_supported_by.append(self.name)

    def __str__(self):
        return f"{self.x1}-{self.x2},{self.y1}-{self.y2},{self.z1}-{self.z2}"


def compute_chain_reaction(map, start):

    items_to_remove = deque()
    items_to_remove.append(start)
    count_removed_items = 0
    while len(items_to_remove) > 0:
        item_to_remove = items_to_remove.popleft()
        count_removed_items+=1
        for i in map[item_to_remove]["supports"]:
            map[i]["is supported"].remove(item_to_remove)
            if len(map[i]["is supported"]) == 0:
                items_to_remove.append(i)
    return count_removed_items-1
    
    

def main():
    parser = argparse.ArgumentParser(description="Day22")
    parser.add_argument("--ex", help="Run on example", action='store_true')
    parser.add_argument("--log", dest="logLevel", choices=['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
                        help="Set the logging level", default="WARNING")
    args = parser.parse_args()
    file = "./src/files/day22.txt"
    if args.ex:
        file = "./src/files/day22ex.txt"

    cubes = []
    with open(file, "r") as f:
        for il, line in enumerate(f.readlines()):
            new_cube = Cube(line, f"{il}")

            cubes.append(new_cube)
    
    cubes.sort(key=lambda x: min(x.z1, x.z2))
    for ix, c in enumerate(cubes):
        c.add_to_ground(cubes[:ix])
    cubes.sort(key=lambda x: [x.z1, x.z2]) 
    only_support = set()
    for c in cubes:
        c.count_supports(cubes)


    reactions_dict = {}
    for c in cubes:
        reactions_dict[c.name] = {"supports":c.supports, "is supported": c.is_supported_by}
        if len(c.is_supported_by) == 1:
            only_support.add(c.is_supported_by[0])

    could_be_removed = len(cubes) - len(only_support)
    print(f"Part 1: {could_be_removed} cubes could be dissintegrated")

    acc = 0
    for i in range(len(cubes)):
        var = copy.deepcopy(reactions_dict)
        ans = compute_chain_reaction(var, f"{i}")
        acc += ans
        #print(f"Removing {i} makes {ans} other bricks fall")
    print(f"Part 2: {acc}")



if __name__ == "__main__":
    main()
